File: solutions/169. Majority Element.cpp
class Solution {
public:
    int majorityElement(vector<int>& nums) {
        int res = nums.at(0);
        int count = 0;
        for (int n : nums) {
            if (!count) {
                res = n;
                ++count;
            }
            else if (n == res) ++count;
            else --count;
        }
        return res;
    }
};

File: solutions/169. Majority Element.rs
impl Solution {
  // Boyer-Moore Voting Algorithm
  pub fn majority_element(nums: Vec<i32>) -> i32 {
      let mut count = 0;
      let mut candidate = nums[0];
      for num in nums {
          if count == 0 {
              candidate = num;
          }
          count += if num == candidate { 1 } else { -1 }
      }
      candidate
  }
}
